From 9e1dad7f825b8e7da943a3568ca79f32da457485 Mon Sep 17 00:00:00 2001
From: Yu Chien Peter Lin <peterlin@andestech.com>
Date: Mon, 22 May 2023 06:47:39 +0000
Subject: [PATCH] rzfive: enable caches

1. The mcache_ctl.CCTL_SUEN should be set so
   ax45mp_iocp_sw_workaround() can return true.
2. Enable L2-cache

These might be done in U-Boot SPL, we will remove
these once U-Boot patches are posted on ML.

Signed-off-by: Yu Chien Peter Lin <peterlin@andestech.com>

---
 platform/generic/renesas/rzfive/rzfive.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/platform/generic/renesas/rzfive/rzfive.c b/platform/generic/renesas/rzfive/rzfive.c
index 0ed5028..fcfed31 100644
--- a/platform/generic/renesas/rzfive/rzfive.c
+++ b/platform/generic/renesas/rzfive/rzfive.c
@@ -6,6 +6,8 @@
 
 #include <andes/andes45_pma.h>
 #include <andes/andes_sbi.h>
+#include <andes/andes45.h>
+#include <sbi/riscv_asm.h>
 #include <platform_override.h>
 #include <sbi/sbi_domain.h>
 #include <sbi_utils/fdt/fdt_helper.h>
@@ -29,8 +31,16 @@ static int renesas_rzfive_final_init(bool cold_boot, const struct fdt_match *mat
 					 array_size(renesas_rzfive_pma_regions));
 }
 
+static void enable_caches(void)
+{
+#define L2C_CTR 0x13400008
+	csr_write(CSR_MCACHE_CTL, 0x183f03ul);
+	*(unsigned long *)L2C_CTR = 0x989;
+}
+
 int renesas_rzfive_early_init(bool cold_boot, const struct fdt_match *match)
 {
+	enable_caches();
 	/*
 	 * Renesas RZ/Five RISC-V SoC has Instruction local memory and
 	 * Data local memory (ILM & DLM) mapped between region 0x30000
